### ISSUE TITLE
Rule proposal: Require/Disallow passing an object to `setState()` if not using previous state

### ISSUE THREAD

**Issue by samvk** [yannickcr/eslint-plugin-react#1836]
`setState(updater)` accepts a function, or an object for the updater if not relying on previous state.

> The first `setState()` argument is an updater function...You may optionally pass an object as the first argument to setState() instead of a function...If the next state depends on the previous state, we recommend using the updater function form." - [docs](https://reactjs.org/docs/react-component.html#setstate)

Proposal: add a rule with the option to either -

- Make `setState(updater)` always take an object if not relying on previous state
- Make `setState(updater)` always take a function, regardless

e.g., if using the former:

```jsx
/* Before: */ this.setState(() => ({quantity: 2}));
/* After: */  this.setState({quantity: 2});
```

---

This is perhap an extension off of `no-access-state-in-setstate`.

**Comment by ljharb** [comment]
I'm not sure why the former would be particularly useful; it'd also make for code churn if you suddenly started or stopped making it rely on previous state.

**Comment by ljharb** [comment]
Given that only one other person has indicated interest in the last 4 years, I'm going to close this. If folks think it's worth reopening, we can.
